app/forkjoin/forkjoin.go

```diff
@@ -0,0 +1,214 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package forkjoin provides an API for "doing work
+// concurrently (fork) and then waiting for the results (join)".
+package forkjoin
+
+import (
+	"context"
+	"sync"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+const (
+	defaultWorkers  = 8
+	defaultInputBuf = 100
+	defaultFailFast = false
+)
+
+// Fork function enqueues the input to be processed asynchronously.
+// Note Fork may block temporarily while the input buffer is full, see WithInputBuffer.
+// Note Fork will panic if called after Join.
+type Fork[I any] func(I)
+
+// Join function closes the input queue and returns the result channel.
+// Note Fork will panic if called after Join.
+// Note Join must only be called once, otherwise panics.
+type Join[I, O any] func() Results[I, O]
+
+// Work defines the work function signature workers will call.
+type Work[I, O any] func(ctx context.Context, input I) (output O, err error)
+
+// Results contains enqueued results.
+type Results[I, O any] <-chan Result[I, O]
+
+// Result contains the input and resulting output from the work function.
+type Result[I, O any] struct {
+	Input  I
+	Output O
+	Err    error
+}
+
+// Flatten blocks and returns all the outputs when all completed and
```

Comment:
nit: Flatten function comment is unclear to me

Reply:
improved